src/java.base/share/classes/java/io/Reader.java

```diff
@@ -196,6 +197,8 @@ public int read(CharBuffer target) throws IOException {
             if (nread > 0)
                 target.position(pos + nread);
         } else {
+            if (target.isReadOnly())
+                throw new ReadOnlyBufferException();
             int len = target.remaining();
```

Comment:
It seems like the other branch of the if should also check for read-only and throw.
Or can the target not have an array if it is readonly.
It might return -1 for a readonly target if the source was used up. The always throw.

Moving the `if (target.isReadOnly())`  to the line 188: would be unambiguous.

Reply:
It can't have an array if it is read-only.
```
Returns:
    true if, and only if, this buffer is backed by an array and is not read-only
```
https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/nio/CharBuffer.html#hasArray()